Fill editing `default_expression` input field with computed default value (eg. `$area`)

volterra79 opened the pull request:
Related to https://github.com/g3w-suite/g3w-client-plugin-editing/issues/35
Related to https://github.com/g3w-suite/g3w-admin/issues/493

Steps:

1. Go to layer "Properties" and click on Attribute form
2. Set a default value as a function (below surface field is set as $area of the feature)

![Screenshot from 2023-03-23 14-21-16](https://user-images.githubusercontent.com/1051694/227218554-5e311024-720c-4434-bfba-0f9b9be8386a.png)

4. Save project and reload a project on G3W-SUITE
5. Go to editing of layer and create a new feature

![Screenshot from 2023-03-23 14-29-27](https://user-images.githubusercontent.com/1051694/227219093-34ad255e-65b5-4ecb-b5a0-0a04f68ad05a.png)

surface field is already filled

To test:

https://github.com/g3w-suite/g3w-admin/ (dev branch)
https://github.com/g3w-suite/g3w-client-plugin-editing/pull/35

volterra79 committed:
Fix default_expression if has a expression that has no field dependencies

volterra79 committed:
Adjust comment

volterra79 committed:
Fill input field in case has a default expression that has an field dependency

volterra79 committed:
change how default expression field is set

volterra79 committed:
Use set default_expression_dependency_fields instead to add new variable

volterra79 committed:
Rename same variable to avoid conflict

Raruto reviewed (commented):
@volterra79 could you add some step-by-step info about how we can to test this?

volterra79 commented:
@Raruto need to some changes and after i write all steps. Thanks

Raruto commented:
ok, I thought it was ready since you asked me for a review.

I move this to draft just to better denote this.

volterra79 committed:
Add field_name param to service data expression and inputservice.js

Raruto commented on src/services/index.js:
```diff
@@ -32,16 +32,16 @@ export {
   RelationsService,
   RouterService,
   TaskService,
-  WorkFlowsStack,
```
Hey @volterra79

> Trailing commas simplify adding and removing items to objects and arrays, since only the lines you are modifying must be touched. Another argument in favor of trailing commas is that it improves the clarity of diffs when an item is added or removed from an object or array

maybe, in a distant future: 👉 [comma-dangle](https://eslint.org/docs/latest/rules/comma-dangle)

volterra79 committed:
https://github.com/g3w-suite/g3w-client/pull/389#discussion_r1146134240

volterra79 committed:
Fix stateValueChanged method. wrong datetimepicker html element set value (date)

volterra79 committed:
In case of expression error evaluation, set field value to default value if set

volterra79 committed:
set datetimedisplayformat and datetimefieldformat as instance properties

Raruto committed:
refactor functions: `handleDefaultExpressionFormInput()` `handleFilterExpressionFormInput()`

Raruto committed:
add new functions

- `_handleFieldWithDefaultExpression(field, default_expression)`
- `_handleFieldWithFilterExpressionfield, filter_expression)`

Raruto committed:
refactor function `_handleFieldWithFilterExpression`

Raruto committed:
☢️ refactor of a setter function

Raruto commented on src/app/gui/inputs/service.js:
```diff
@@ -29,26 +29,46 @@ proto.getValue = function() {
-      //check if we can state.check get_default_value from input.options.default is set
-      if ( this.state.get_default_value && typeof this.state.input.options.default !== "undefined" && this.state.input.options.default !== null) {
-        this.state.value = this.state.input.options.default;
-        this.state.value_from_default_value = true;
-      }
+  
+  // skip when ??
+  if (!(null === value || undefined === value)) {
+    return;
+  }
+
+  const options = this.state.input.options;
+
+  // when .. ??
+  if (Array.isArray(options)) {
+    if (options[0].default) {
+      this.state.value = options[0].default;
+    } else if(values.length && Array.isArray(options.values)) {
+      this.state.value = ( options.values[0] && options.values[0].value || options.values[0]);
     }
   }
+
+  const FIND_ME_A_NAME = (
+    !Array.isArray(options) &&
+    this.state.get_default_value &&
+    undefined !== options.default &&
+    null !== options.default
+  );
+
+  // check if we can state.check get_default_value from input.options.default is set
+  if (FIND_ME_A_NAME && undefined === options.default_expression) {
+      this.state.value = options.default;
+  }
+
+  // when .. ??
+  if(FIND_ME_A_NAME) {
+    this.state.value_from_default_value = true;
+  }
+
 };
```
remember to double check, and try to simplify further

volterra79 committed:
change _handleFieldWithFilterExpression to _handleFieldWithDefaultExpression

Raruto committed:
missing variable name

Raruto commented on src/app/gui/inputs/service.js:
```diff
@@ -29,32 +29,46 @@ proto.getValue = function() {
+    if (options[0].default) {
+      this.state.value = options[0].default;
+    } else if(values.length && Array.isArray(options.values)) {
+      this.state.value = ( options.values[0] && options.values[0].value || options.values[0]);
```
there is something strange in this condition

volterra79 committed:
set name to FIND_ME_A_NAME and refactor method

volterra79 committed:
Fix wrong default_expression_dependency_fields set reference variable name. Add comment

Raruto committed:
fix `typeofdefault_value`

Raruto committed:
refactor function `core/layers/tablelayer.js::getFieldsWithValues()`

Raruto committed:
Update InputDateTimePicker.vue

Raruto committed:
slim down `services\data-expression.js`

Raruto committed:
defaul values

Raruto committed:
rename duplicate classes: `Routerservice`

- add new class `DataService`
- add new class `HistoryService`

Raruto commented:
@volterra79 am i missing something?

![image](https://user-images.githubusercontent.com/9614886/227538338-5bd66394-173c-4507-b49f-9454626f3222.png)

![image](https://user-images.githubusercontent.com/9614886/227538136-5fea0075-0146-497d-8c04-b3228036a3db.png)

![image](https://user-images.githubusercontent.com/9614886/227541318-9ac42e9b-82e8-473d-a79d-0a9250ef96d8.png)

volterra79 commented:
@Raruto sorry. To test it you need to use https://github.com/g3w-suite/g3w-client-plugin-editing/pull/35

Raruto commented:
It could be related to a my previous edit in here? [src/services/data-expression.js#L64-65](https://github.com/g3w-suite/g3w-client/pull/389/files#diff-65f1aa97440274a227ae6c4424f993de2b5a6c4d0716fff3d4688afb063f1e9aR64-R65)

---

I'm testing with:

- g3w-admin: __3.5.dev-20230324131205__
- [g3w-suite-demo-projects/editing-default-expression-$area.qgs ](https://github.com/g3w-suite/g3w-suite-demo-projects/blob/703f4d7ff47daabd6378cbeb2896a9ff0ad00eaa/projects/editing-default-expression-%24area.qgs)
- https://github.com/g3w-suite/g3w-client-plugin-editing/pull/35

and i get the following immediately after drawing a new feature:

![image](https://user-images.githubusercontent.com/9614886/227551982-6702a9b5-81d1-404c-81e2-9e13dc8caa8c.png)

## Request

```url
XHR POST http://localhost:8000/api/expression_eval/1/
[HTTP/1.1 400 Bad Request 152 ms]

{
    "field_name": "surface",
    "layer_id": "buildings_2f43dc1d_6725_42d2_a09b_dd446220104a",
    "qgs_layer_id": "buildings_2f43dc1d_6725_42d2_a09b_dd446220104a",
    "form_data": {
        "type": "Feature",
        "geometry": {
            "type": "MultiPolygon",
            "coordinates": [
                [
                    [
                        [1251381.3475698456, 5433508.099718014],
                        [1251452.1733660277, 5433536.602782331],
                        [1251473.766596571, 5433509.395311846],
                        [1251459.0831998016, 5433484.347164417],
                        [1251388.6892682302, 5433488.665810525],
                        [1251381.3475698456, 5433508.099718014]
                    ]
                ]
            ]
        },
        "properties": {
            "fid": null,
            "name": null,
            "address": null,
            "year": null,
            "photo": null,
            "link": null,
            "form": null,
            "high": null,
            "volume": null,
            "surface": null,
            "architectural_barriers": null,
            "date_barriers": null,
            "safety_exits": null,
            "date_exits": null,
            "fire_system": null,
            "date_fire": null,
            "type": null,
            "zone": null,
            "attachment": null,
            "subtype": null,
            "user": null,
            "description": null
        },
        "id": "_new_217_1679667526404"
    },
    "formatter": 1,
    "expression": "$area"
}
```

## Response

```url
XHR POST http://localhost:8000/api/expression_eval/1/
[HTTP/1.1 400 Bad Request 152ms]

{
    "result": false,
    "error": {
        "code": "servererror",
        "message": "A error server is occured!",
        "data": ""
    }
}
```

volterra79 commented:
@Raruto need  to test with **g3w-admin:dev**

Raruto commented:
Testing also with just: https://github.com/g3w-suite/g3w-admin/commit/31108747e98e8e96b92877bd05c0547dae2ddb39

That means, without runnin `npm run dev` and with a clean docker `config\g3w-suite\overrides` folder, and i get the same result (the strange thing is that when looking at the server log there's no error or fatal message..)

volterra79 committed:
set cont or let variable instead global variable

volterra79 commented:
> Testing also with just: [g3w-suite/g3w-admin@3110874](https://github.com/g3w-suite/g3w-admin/commit/31108747e98e8e96b92877bd05c0547dae2ddb39)
> 
> That means, without runnin `npm run dev` and with a clean docker `config\g3w-suite\overrides` folder, and i get the same result (the strange thing is that when looking at the server log there's no error or fatal message..)

@Raruto very strange. 
https://dev.g3wsuite.it/en/map/g3w-suite-demo/qdjango/97/ there is working example

@wlorenzetti what do you think about?

wlorenzetti commented:
It's very strange, what does the docker log say?

volterra79 committed:
Remove formatter parameter !! 1. In case of formatter = 0 set 1 wrong.

volterra79 committed:
change expr to params

Raruto commented:
After https://github.com/g3w-suite/g3w-client/pull/389/commits/25c0faebe74930589b0d34babdef62d1eeb6af63 the `$area` surface seems to be calculated automatically.

However, now I also see this other strange request:

![image](https://user-images.githubusercontent.com/9614886/227860236-4a804c76-174f-4a09-9ad3-17ee450338a9.png)

---

## Request

```url
XHR POST http://localhost:8000/api/expression_eval/1/
[HTTP/1.1 400 Bad Request 187ms]

{
    "field_name": "user",
    "layer_id": "buildings_2f43dc1d_6725_42d2_a09b_dd446220104a",
    "qgs_layer_id": "buildings_2f43dc1d_6725_42d2_a09b_dd446220104a",
    "form_data": {
        "type": "Feature",
        "geometry": {
            "type": "MultiPolygon",
            "coordinates": [
                [
                    [
                        [1251546.888991395, 5433738.981618131],
                        [1251839.7356094709, 5433802.431718714],
                        [1251930.0299833775, 5433670.65074058],
                        [1251817.7721131153, 5433555.952481833],
                        [1251573.7332647187, 5433607.200639997],
                        [1251546.888991395, 5433738.981618131]
                    ]
                ]
            ]
        },
        "properties": {
            "fid": null,
            "name": null,
            "address": null,
            "year": null,
            "photo": null,
            "link": null,
            "form": null,
            "high": null,
            "volume": null,
            "surface": null,
            "architectural_barriers": null,
            "date_barriers": null,
            "safety_exits": null,
            "date_exits": null,
            "fire_system": null,
            "date_fire": null,
            "type": null,
            "zone": null,
            "attachment": null,
            "subtype": null,
            "user": null,
            "description": null
        },
        "id": "_new_95_1679898433809"
    },
    "formatter": 0,
    "expression": "test"
}

```

## Response

```url
XHR POST http: //localhost:8000/api/expression_eval/1/
[HTTP/1.1 400 Bad Request 123 ms]

{
    "result": false,
    "error": {
        "code": "servererror",
        "message": "A error server is occured!",
        "data": "Field 'test' not found"
    }
}
```

If I remember correctly `"test"` was the name of the temporary layer (and not an expression, which now seems to have a lock) I tried to create last time I've tried this pull request.

Anyway, I'll try again starting from a clean project 🧹🗑️

Raruto commented:
> ![image](https://user-images.githubusercontent.com/9614886/227860236-4a804c76-174f-4a09-9ad3-17ee450338a9.png)

Seems like a bug related to all demo projects we have created:

![image](https://user-images.githubusercontent.com/9614886/227868105-6fe5ded3-b09a-4f64-ac19-07ef43b0d429.png)

**Reference:**

- [g3w-suite-demo-projects/projects/editing-default-expression-$area.qgs#L2095](https://github.com/g3w-suite/g3w-suite-demo-projects/blob/703f4d7ff47daabd6378cbeb2896a9ff0ad00eaa/projects/editing-default-expression-%24area.qgs#L2095)
- [G3WSUITE_3.5.zip](https://drive.google.com/file/d/1x66-47m4QHt2fzkAuphVZp9B_QmVSFIY/view?usp=share_link)

volterra79 commented:
> > ![image](https://user-images.githubusercontent.com/9614886/227860236-4a804c76-174f-4a09-9ad3-17ee450338a9.png)
> 
> Seems like a bug related to all demo projects we have created:
> 
> ![image](https://user-images.githubusercontent.com/9614886/227868105-6fe5ded3-b09a-4f64-ac19-07ef43b0d429.png)
> 
> **Reference:**
> 
> * [g3w-suite-demo-projects/projects/editing-default-expression-$area.qgs#L2095](https://github.com/g3w-suite/g3w-suite-demo-projects/blob/703f4d7ff47daabd6378cbeb2896a9ff0ad00eaa/projects/editing-default-expression-%24area.qgs#L2095)
> * [G3WSUITE_3.5.zip](https://drive.google.com/file/d/1x66-47m4QHt2fzkAuphVZp9B_QmVSFIY/view?usp=share_link)

@Raruto test is consider a field that not exist in layer

volterra79 committed:
Fix position marker point on click map position coordinate

volterra79 committed:
Call widget change only if value is changed

volterra79 committed:
In case of datetimepicker input type need to be compare in Upper case the value because can happen that based on location date can be for example 30 gen 2023 and 30 Gen 2023

volterra79 committed:
Create saveDefaultExpressionFieldsNotDependencies method and default_expression_no_fields_dependencies to handle fields that have defualt_expression but not related to dependency fields

volterra79 committed:
Add input datetime i18n translation

volterra79 committed:
Remove wrong rule

volterra79 committed:
remove widgetMixins dependence in favor of own more control flow of state.value change

volterra79 committed:
Fix conflict with input state.value watch

volterra79 committed:
Remove click event on info query result. info.action was removed

volterra79 committed:
handle default_expression_fields_on_update on save if apply_on_update
